internal/transports/p2p/peer/peer.go

```diff
@@ -501,6 +506,15 @@ func (p *Peer) handleHeadersMsg(msg *wire.MsgHeaders) {
 
 	if headersReceived == 0 {
 		p.log.Debug().Msgf("received only existing headers from peer: %s", p)
+		// If we're synced with all checkpoints and received only existing headers
+		// then it likely means that we're synced with tip of chain, so now we can
+		// request the peer to send us headers directly instead of inv messages.
+		//
+		// If we do this while still syncing, we will get a lot of orphaned blocks,
+		// because the new headers sent by peer will not have parents yet.
+		if p.syncedCheckpoints {
```

Comment:
Maybe instead we should consider a flag (or method) that will tell that we're synced with peers tip. 
And how we can assume that we're synced?
- with Version message we're receiving start_height which means what is the height of this peer at the moment of sending a Version message
- everytime we receive INV we know what is the highest hash of this peer.

Having those information we could assume that we're synced when:
- hash from INV is nil and our tip is at the start_height
- hash from INV is not nil and we have such a hash in db

Then we could use that info we could switch to sendheaders as soon as we reach the tip from the peer.

Reply:
⚠️ We should probably also take into account the time when we last time received INV message, if it is more then  for example 1 or 2 minute old, then we should wait for the next block. This way we will be more safe to not omit some header because of the concurency

Reply:
This is only suggestio, fill free to do it any way that would be right.
ℹ️ I would probably go with solution like that:
1. add 2 fields to peer: latestHash and latestHeight
2. when receiving version set latestHeight from version
3. when receiving INV, set latestHash from it and remove latestHeight (set it to nil)
4. when we save header then we can check if it's hash is equal to latestHash and if so, then set latesHeight to the header height
5. when checking if we're synced, check if tip is eq to latestHeight - we could also make this kinda cacheable and set the flag (synced) at the peer in step 4 instead.

Reply:
That's a great idea. I finally managed to come up with a simple and elegant (IMHO) solution, based on your suggested flow, but also adding couple of safeguards, checks and a mutex to be thread-safe :)

Please check if the solution now is satisfactory :)